### DISCUSSION TITLE
Using getInitialProps in _app.js with an external API call

### DISCUSSION THREAD

**Discussion by MikeWillis** [vercel/next.js#27519] [Help]
TL;DR: Is it possible to have an external API call in a `getInitialProps` in `_app.js`, and ONLY make that call the first time the app is loaded?

Details:
I'm working on a NextJS app that will function as the UI for an ecommerce site. I was thinking it would be nice to not have to rebuild the app every time the main navigation changes. Not that it changes frequently, but still, categories are added and removed now and then.

So I set up up the navigation so it can be retrieved as JSON from an API call, and then I set up `_app.js` to use `getInitialProps` to make the call. That worked fine. I also set up that `getInitialProps` to run any `getInitialProps` of the page being rendered. All is good, I get a fully hydrated page including the navigation.

However, I noticed today that the call to the API that pulls in the navigation is being made on every route change. In hindsight it seems to make sense, because the `getInitialProps` in `_app.js` has to run on every route change so it can run the current page's `getInitialProps`.

So I'm thinking I should give up and just code the navigation into the app, and do a rebuild whenever the nav changes. It certainly doesn't make sense to re-fetch it on every route change.

Just wondering if anyone else had a similar issue and came up with a workaround? I mean a navigation is one thing that can easily be coded into the app, but what if there was some content that had to be pull data in from an external API on the first render - content that you wanted to be rendered server-side and therefore the API call had to live in a `getInitialProps` (or `getServerSideProps`)?
